### ISSUE TITLE
Unexpected diffs on network resources with v2 alpha

### ISSUE THREAD

**Issue by mnlumi** [pulumi/pulumi-azure-native#2475]
### What happened?

When working with Pulumi templates and testing out 2.0.0-alpha.1, there are diffs in network resources. ~that do not occur in Python or Typescript~. **Edit**: Upon further testing, these are also occurring in Typescript and Python.
![Screenshot 2023-05-26 at 11 38 03 AM](https://github.com/pulumi/pulumi-azure-native/assets/118926824/7f0518b1-a6e9-4bc6-b918-54a9ec151017)
![Screenshot 2023-05-26 at 11 37 59 AM](https://github.com/pulumi/pulumi-azure-native/assets/118926824/d1888628-e164-44b2-a6a9-281c7206a0c7)
![Screenshot 2023-05-26 at 11 37 54 AM](https://github.com/pulumi/pulumi-azure-native/assets/118926824/688ad636-38cd-4149-ba3f-9f6553f9f338)
![Screenshot 2023-05-26 at 11 37 49 AM](https://github.com/pulumi/pulumi-azure-native/assets/118926824/2c317fbc-5080-48b9-8c71-4a3bb5a33feb)

### Expected Behavior

I expect ~consistent behavior across languages and~ no diffs for these resources when upgrading from v1.102.0 to 2.0.0-alpha.1.

### Steps to reproduce

Using any of the templates listed, run the program first with v1.102.0 to create the resources, then upgrade the provider to 2.0.0-alpha.1 and run `pulumi up`.

[container-azure-csharp](https://github.com/pulumi/templates/tree/master/container-azure-csharp) 
[kubernetes-azure-csharp](https://github.com/pulumi/templates/tree/master/kubernetes-azure-csharp) 
[serverless-azure-csharp](https://github.com/pulumi/templates/tree/master/serverless-azure-csharp) 
[vm-azure-csharp](https://github.com/pulumi/templates/tree/master/vm-azure-csharp) 

### Output of `pulumi about`

```
CLI
Version      3.68.0
Go Version   go1.20.4
Go Compiler  gc

Plugins
NAME           VERSION
azure-native   2.0.0-alpha.1
dotnet         unknown
synced-folder  0.10.2

Host
OS       darwin
Version  13.0.1
Arch     arm64

This project is written in dotnet: executable='/usr/local/share/dotnet/dotnet' version='7.0.302'
```

### Contributing

Vote on this issue by adding a 👍 reaction. 
To contribute a fix for this issue, leave a comment (and link to your pull request, if you've opened one already).

**Comment by mikhailshilkov** [comment]
I got the same diff with `container-azure-typescript`. I suspect it's expected due to API version bumps.

<img width="1768" alt="image" src="https://github.com/pulumi/pulumi-azure-native/assets/1454008/bfe98a53-5639-4279-b293-c8c7f284becd">

**Comment by mnlumi** [comment]
After further testing, I am seeing the diffs pending in python as well, so this is not language specific. I've updated the issue title and description.
![Screenshot 2023-05-26 at 11 37 05 AM](https://github.com/pulumi/pulumi-azure-native/assets/118926824/32a04919-34df-4f8d-b02e-b6f0d32671a7)

**Comment by mikhailshilkov** [comment]
That's good to hear, one question answers.

As for the changes, I believe they are according to our current design for v2.

Let's take one change as an example, say `azure-native:containerregistry:Registry` added `publicNetworkAccess     : "Enabled"`. Here are the steps to investigate:

1. Check the old API version, it's [`2019-05-01`](https://github.com/pulumi/pulumi-azure-native/blob/876c1b263bdaa964e9fe1063c7528c7cd05458f9/sdk/dotnet/ContainerRegistry/Registry.cs#L14).
2. Check the new API version, it's [`2022-12-01`](https://github.com/pulumi/pulumi-azure-native/blob/27b5d7a6d2430f7f68acb15bace3ef3a84db8aec/sdk/dotnet/ContainerRegistry/Registry.cs#L14).
3. Check if the old API version has a property called `publicNetworkAccess`: [it does not](https://github.com/Azure/azure-rest-api-specs/blob/main/specification/containerregistry/resource-manager/Microsoft.ContainerRegistry/stable/2019-05-01/containerregistry.json).
4. Check if the new API version has a property called `publicNetworkAccess`: [it does](https://github.com/Azure/azure-rest-api-specs/blob/main/specification/containerregistry/resource-manager/Microsoft.ContainerRegistry/stable/2022-12-01/containerregistry.json#L2685C10-L2687).
5. It also has a default value set to `Enabled`. That's why the diff is shown when migrating.

@mnlumi Should we close this as by-design? Or what other actions would you like us to do?

**Comment by mnlumi** [comment]
Yes, let's close this as by-design. Thanks!

**Comment by kpitzen** [comment]
Reopening due to this actually not being the behavior on other providers. Context to follow.

**Comment by danielrbradley** [comment]
As discussed in this comment, I'm also of the opinion that this is currently fuctioning as designed: https://github.com/pulumi/pulumi-azure-native/pull/2506#issuecomment-1584396048

1. If the version (1) of a resource has set a default pricing tier of "A".
2. If we upgrade to version (2) of the resource where the new API version sets the default pricing tier to "B" (perhaps more expensive).
3. If the user hasn't set the pricing tier explicitly, would they not want to be notified that their pricing tier is changing?

**Comment by mikhailshilkov** [comment]
@danielrbradley Yes, we want to see a diff in the situation that you describe.

What we want to avoid is a diff when

1. Version (1) has no property for the pricing tier, so whatever default A is used behind the scenes.
2. Version (2) has a new property for the pricing tier which sets it to A.

In theory, 2 could set it to B, but I don't think that's a reasonable API design that exists in reality.

**Comment by thomas11** [comment]
The just merged #2506 solves the issue for top-level properties. We still need to recurse into nested and array properties to handle their new defaults.

**Comment by mikhailshilkov** [comment]
We think we resolved all the main scenarios. If diffs pop up again, we'll track them in a separate issue.
